File: services/nlp/app/services/extract/epub.py
"""
EPUB parsing and text extraction.

Python port of the PHP ``EpubParserService``
(``src/Modules/Book/Application/Services/EpubParserService.php``). Reads an EPUB
(either fetched from a URL via the SSRF-safe util or supplied as raw bytes),
walks its document items in spine order, converts each chapter's HTML to plain
text, and returns normalised metadata + chapters + the joined full text.

``ebooklib`` and ``bs4`` (BeautifulSoup) are lazy-imported inside the parsing
function so this router still mounts when they are not installed.
"""

from __future__ import annotations

import os
import re
import tempfile

from fastapi import HTTPException

from app.services.http.safe_fetch import FetchError, safe_get

# Cap total extracted text — a guard against absurdly large / malicious books.
_MAX_TEXT_CHARS = 5_000_000
# Below this many words the book is almost certainly image-only (picture book).
_MIN_WORDS = 30
# A chapter title candidate must be short and contain no period.
_MAX_TITLE_LEN = 100
# Documents yielding less than this many characters are treated as nav/toc noise.
_MIN_DOC_CHARS = 30

_NAV_HINTS = ("nav", "toc")


async def extract_from_url(url: str) -> dict:
    """Fetch an EPUB from ``url`` and parse it.

    Routes through the SSRF-safe fetch util with EPUB-appropriate caps, then
    delegates to :func:`parse_bytes`.
    """
    try:
        result = await safe_get(
            url,
            timeout=30.0,
            max_bytes=20_000_000,
            accept="application/epub+zip",
        )
    except FetchError as exc:
        raise HTTPException(exc.status_code, exc.message) from exc

    return parse_bytes(result.content, result.url)


def _normalize_whitespace(text: str) -> str:
    """Collapse whitespace while preserving paragraph (double-newline) breaks."""
    text = text.replace("\r", "\n")
    # Collapse runs of spaces/tabs.
    text = re.sub(r"[ \t]+", " ", text)
    # Trim each line.
    lines = [line.strip() for line in text.split("\n")]
    text = "\n".join(lines)
    # Collapse 3+ newlines into a paragraph break.
    text = re.sub(r"\n{3,}", "\n\n", text)
    return text.strip()


def _is_navigation_doc(filename: str, text: str) -> bool:
    """Skip nav/toc documents: name hints or near-empty text content."""
    lowered = (filename or "").lower()
    if any(hint in lowered for hint in _NAV_HINTS):
        return True
    return len(text.strip()) < _MIN_DOC_CHARS


def _chapter_title(text: str, num: int) -> str:
    """First non-empty line if it looks like a title, else ``Chapter N``.

    Mirrors the PHP ``extractTitleFromContent``: a candidate title must be
    shorter than 100 chars and contain no period.
    """
    for line in text.split("\n"):
        first = line.strip()
        if first:
            if len(first) < _MAX_TITLE_LEN and "." not in first:
                return first
            break
    return f"Chapter {num}"


def _first_metadata(book, namespace: str, name: str) -> str | None:
    """Return the first Dublin Core metadata value, or None when absent."""
    try:
        entries = book.get_metadata(namespace, name)
    except Exception:  # noqa: BLE001 - tolerate malformed metadata
        return None
    if not entries:
        return None
    value = entries[0][0] if isinstance(entries[0], (tuple, list)) else entries[0]
    if value is None:
        return None
    value = str(value).strip()
    return value or None


def parse_bytes(data: bytes, source_uri: str = "") -> dict:
    """Parse EPUB ``data`` (in-memory bytes) into metadata, chapters and text.

    ``ebooklib`` requires a filesystem path, so the bytes are written to a
    temporary ``.epub`` file, read, and unlinked (try/finally). Returns:

    ``{"metadata": {...}, "chapters": [...], "text": ...}`` or
    ``{"error": ...}`` when the book has too little readable text.
    """
    try:
        import ebooklib
        from ebooklib import epub
    except ImportError as exc:
        raise HTTPException(
            503,
            "EPUB extraction unavailable: ebooklib is not installed",
        ) from exc

    try:
        from bs4 import BeautifulSoup
    except ImportError as exc:
        raise HTTPException(
            503,
            "EPUB extraction unavailable: bs4 is not installed",
        ) from exc

    tmp_path = ""
    try:
        with tempfile.NamedTemporaryFile(suffix=".epub", delete=False) as tmp:
            tmp.write(data)
            tmp_path = tmp.name
        try:
            book = epub.read_epub(tmp_path)
        except Exception as exc:  # noqa: BLE001 - corrupt / non-EPUB input
            raise HTTPException(422, f"Could not read EPUB file: {exc}") from exc
    finally:
        if tmp_path and os.path.exists(tmp_path):
            os.remove(tmp_path)

    metadata = {
        "title": _first_metadata(book, "DC", "title") or "Unknown Title",
        "author": _first_metadata(book, "DC", "creator"),
        "description": _first_metadata(book, "DC", "description"),
        "language": _first_metadata(book, "DC", "language"),
        "sourceUri": source_uri,
    }

    chapters: list[dict] = []
    total_chars = 0
    chapter_num = 1

    for item in book.get_items_of_type(ebooklib.ITEM_DOCUMENT):
        filename = item.get_name() or ""
        try:
            raw = item.get_content()
        except Exception:  # noqa: BLE001 - skip unreadable items
            continue
        soup = BeautifulSoup(raw, "html.parser")
        text = _normalize_whitespace(soup.get_text("\n"))

        if _is_navigation_doc(filename, text):
            continue
        if not text:
            continue

        # Enforce the total-text cap (zip-bomb / absurd-content guard).
        if total_chars + len(text) > _MAX_TEXT_CHARS:
            text = text[: max(0, _MAX_TEXT_CHARS - total_chars)]
        total_chars += len(text)

        chapters.append(
            {
                "num": chapter_num,
                "title": _chapter_title(text, chapter_num),
                "content": text,
            }
        )
        chapter_num += 1

        if total_chars >= _MAX_TEXT_CHARS:
            break

    full_text = "\n\n".join(ch["content"] for ch in chapters)

    if len(full_text.split()) < _MIN_WORDS:
        return {
            "error": "This book has too little readable text — "
            "it is likely an image-only picture book."
        }

    return {"metadata": metadata, "chapters": chapters, "text": full_text}

File: src/frontend/js/shared/offline/local/content/epub.ts
/**
 * EPUB parsing + plain-text extraction, ported to run client-side.
 *
 * Faithful port of the server's EPUB import. The HTML→text cleaning mirrors
 * `EpubParserService::cleanHtmlContent` (PHP, the canonical production path)
 * regex-for-regex; the spine walk, nav-doc skipping, title fallback and the
 * min-words / max-chars guards mirror `services/nlp/.../extract/epub.py` (the
 * URL-fetch path this client most resembles). Unzipping uses fflate — a tiny
 * pure-JS inflate — so it runs on any WebView with no native plugin and no
 * dependency on a modern `DecompressionStream`; the OPF / container XML is read
 * with the platform `DOMParser`.
 *
 * Like the Python service, all chapters are joined with a blank line into one
 * text body: the on-device model has no multi-chapter Book entity (offline
 * texts are standalone), so the 65 KB-per-text chapter splitting the server
 * does for its MySQL `TEXT` column is unnecessary here.
 *
 * @license Unlicense <http://unlicense.org/>
 */

import { unzipSync, type Unzipped } from 'fflate';
import { corsFreeGetBytes } from './native-http';

// Guards mirroring the Python service (`extract/epub.py`).
const MAX_TEXT_CHARS = 5_000_000; // total extracted-text cap (zip-bomb guard)
const MIN_WORDS = 30; // below this the book is almost certainly image-only
const MIN_DOC_CHARS = 30; // documents shorter than this are nav/toc noise
const MAX_ENTRIES = 2000; // refuse archives with implausibly many files

/** The plain-text book extracted from an EPUB. */
export interface ParsedEpub {
  title: string;
  /** Every chapter joined (kept for coverage previews). */
  text: string;
  /** The spine's chapters in order — the per-chapter import splits on these. */
  chapters: string[];
}

/** A per-chapter text payload to create from a parsed EPUB. */
export interface EpubChapterText {
  title: string;
  text: string;
  tags?: string[];
}

/**
 * Turn a parsed EPUB's chapters into per-chapter text payloads — the settled
 * on-device book model (Option A): an EPUB becomes one normal text per chapter,
 * grouped in the library by a shared tag (the book title), with no persistent
 * book entity. A single-chapter book stays one untagged text.
 */
export function epubChapterTexts(bookTitle: string, chapters: string[]): EpubChapterText[] {
  const nonEmpty = chapters.filter((c) => c.trim() !== '');
  const total = nonEmpty.length;
  if (total <= 1) {
    return [{ title: bookTitle, text: nonEmpty[0] ?? chapters.join('\n\n') }];
  }
  return nonEmpty.map((text, index) => ({
    title: `${bookTitle} — ${index + 1}/${total}`,
    text,
    tags: [bookTitle],
  }));
}

// --- HTML → plain text (port of EpubParserService::cleanHtmlContent) ---------

let decoderEl: HTMLTextAreaElement | null = null;

/**
 * Decode HTML entities, equivalent to PHP's `html_entity_decode(ENT_HTML5)`.
 * Uses a detached `<textarea>` (the same trick as the GDL client) when a DOM is
 * present, with a minimal manual fallback for headless contexts. `&nbsp;`
 * decodes to U+00A0 (as PHP does), which the later space-collapsing step leaves
 * intact — matching the server byte-for-byte.
 */
function decodeEntities(text: string): string {
  if (!text || text.indexOf('&') === -1) {
    return text;
  }
  if (typeof document !== 'undefined') {
    if (!decoderEl) {
      decoderEl = document.createElement('textarea');
    }
    decoderEl.innerHTML = text;
    return decoderEl.value;
  }
  return text
    .replace(/&nbsp;/g, ' ')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#0*39;/g, "'")
    .replace(/&apos;/g, "'")
    .replace(/&#x?[0-9a-fA-F]+;/g, (m) => {
      const hex = /^&#x/i.test(m);
      const code = parseInt(m.slice(hex ? 3 : 2, -1), hex ? 16 : 10);
      return Number.isNaN(code) ? m : String.fromCodePoint(code);
    })
    .replace(/&amp;/g, '&');
}

/**
 * Convert one chapter's (X)HTML to plain text, preserving paragraph structure
 * as blank lines. Mirrors `EpubParserService::cleanHtmlContent` step for step
 * (`[\s\S]` stands in for PHP's `s`/dotall flag, the convention used elsewhere
 * in this folder).
 */
export function cleanHtmlContent(html: string): string {
  // Remove scripts and styles entirely.
  let text = html.replace(/<script\b[^>]*>[\s\S]*?<\/script>/gi, '');
  text = text.replace(/<style\b[^>]*>[\s\S]*?<\/style>/gi, '');
  // Paragraphs and divs -> blank line.
  text = text.replace(/<\/?(?:p|div)[^>]*>/gi, '\n\n');
  // Line breaks -> newline.
  text = text.replace(/<br\s*\/?>/gi, '\n');
  // Headings (open + close) -> blank line.
  text = text.replace(/<h[1-6][^>]*>/gi, '\n\n');
  text = text.replace(/<\/h[1-6]>/gi, '\n\n');
  // List items -> bullet.
  text = text.replace(/<li[^>]*>/gi, '\n- ');
  // Strip HTML comments, then all remaining tags (strip_tags removes both).
  text = text.replace(/<!--[\s\S]*?-->/g, '');
  text = text.replace(/<[^>]*>/g, '');
  // Decode HTML entities.
  text = decodeEntities(text);
  // Collapse runs of spaces/tabs/CRs to a single space.
  text = text.replace(/[ \t\r]+/g, ' ');
  // Blank-line runs -> a single blank line.
  text = text.replace(/\n\s*\n/g, '\n\n');
  // Trim each line.
  text = text
    .split('\n')
    .map((line) => line.trim())
    .join('\n');
  // Cap at a paragraph break.
  text = text.replace(/\n{3,}/g, '\n\n');
  return text.trim();
}

// --- ZIP / OPF / spine -------------------------------------------------------

const utf8 = new TextDecoder('utf-8');

function parseXml(bytes: Uint8Array): Document {
  return new DOMParser().parseFromString(utf8.decode(bytes), 'application/xml');
}

/** Elements by local name, ignoring XML namespace prefixes. */
function byLocalName(root: Document, name: string): Element[] {
  return Array.from(root.getElementsByTagNameNS('*', name));
}

function dirName(path: string): string {
  const idx = path.lastIndexOf('/');
  return idx === -1 ? '' : path.slice(0, idx);
}

function baseName(path: string): string {
  const idx = path.lastIndexOf('/');
  return idx === -1 ? path : path.slice(idx + 1);
}

/** Resolve a manifest href (relative to the OPF dir) to a ZIP entry path. */
function resolvePath(baseDir: string, href: string): string {
  let h = href.split('#')[0].split('?')[0];
  try {
    h = decodeURIComponent(h);
  } catch {
    // Keep the raw href if it isn't valid percent-encoding.
  }
  if (h.startsWith('/')) {
    h = h.slice(1);
  }
  const parts = (baseDir ? baseDir.split('/') : []).concat(h.split('/'));
  const stack: string[] = [];
  for (const part of parts) {
    if (part === '' || part === '.') {
      continue;
    }
    if (part === '..') {
      stack.pop();
      continue;
    }
    stack.push(part);
  }
  return stack.join('/');
}

/** Read `META-INF/container.xml` to find the OPF package path ('' if absent). */
function findOpfPath(zip: Unzipped): string {
  const container = zip['META-INF/container.xml'];
  if (!container) {
    return '';
  }
  for (const rootfile of byLocalName(parseXml(container), 'rootfile')) {
    const fullPath = rootfile.getAttribute('full-path');
    if (fullPath) {
      return fullPath;
    }
  }
  return '';
}

/** Parse the OPF: the book title and the spine's content documents, in order. */
function parseOpf(zip: Unzipped, opfPath: string): { title: string; docs: string[] } {
  const doc = parseXml(zip[opfPath]);
  const baseDir = dirName(opfPath);

  let title = '';
  for (const el of byLocalName(doc, 'title')) {
    const value = (el.textContent ?? '').trim();
    if (value) {
      title = value;
      break;
    }
  }

  const hrefById = new Map<string, string>();
  for (const item of byLocalName(doc, 'item')) {
    const id = item.getAttribute('id');
    const href = item.getAttribute('href');
    if (id && href) {
      hrefById.set(id, resolvePath(baseDir, href));
    }
  }

  const docs: string[] = [];
  for (const itemref of byLocalName(doc, 'itemref')) {
    const idref = itemref.getAttribute('idref');
    if (!idref) {
      continue;
    }
    const path = hrefById.get(idref);
    if (path && zip[path]) {
      docs.push(path);
    }
  }

  return { title, docs };
}

/** Skip nav/toc documents: name hints, or near-empty cleaned text. */
function isNavigationDoc(path: string, cleanedText: string): boolean {
  const name = baseName(path).toLowerCase();
  if (name.includes('nav') || name.includes('toc')) {
    return true;
  }
  return cleanedText.trim().length < MIN_DOC_CHARS;
}

function countWords(text: string): number {
  const trimmed = text.trim();
  return trimmed === '' ? 0 : trimmed.split(/\s+/).length;
}

/**
 * Parse EPUB bytes into a single plain-text book. Returns `{ error }` when the
 * archive is unreadable or carries too little text to be worth importing.
 */
export function parseEpub(bytes: Uint8Array): ParsedEpub | { error: string } {
  let zip: Unzipped;
  try {
    zip = unzipSync(bytes);
  } catch {
    return { error: 'Could not read the EPUB (not a valid ZIP archive).' };
  }

  const names = Object.keys(zip);
  if (names.length > MAX_ENTRIES) {
    return { error: 'This EPUB has too many files.' };
  }

  const opfPath = findOpfPath(zip);
  let title = '';
  let docs: string[] = [];
  if (opfPath && zip[opfPath]) {
    const parsed = parseOpf(zip, opfPath);
    title = parsed.title;
    docs = parsed.docs;
  }
  // Fallback (no container / empty spine): every (x)html doc in path order.
  if (docs.length === 0) {
    docs = names.filter((n) => /\.x?html?$/i.test(n)).sort();
  }

  const chapters: string[] = [];
  let totalChars = 0;
  for (const path of docs) {
    const raw = zip[path];
    if (!raw) {
      continue;
    }
    let text = cleanHtmlContent(utf8.decode(raw));
    if (isNavigationDoc(path, text)) {
      continue;
    }
    if (!text) {
      continue;
    }
    if (totalChars + text.length > MAX_TEXT_CHARS) {
      text = text.slice(0, Math.max(0, MAX_TEXT_CHARS - totalChars));
    }
    totalChars += text.length;
    chapters.push(text);
    if (totalChars >= MAX_TEXT_CHARS) {
      break;
    }
  }

  const fullText = chapters.join('\n\n');
  if (countWords(fullText) < MIN_WORDS) {
    return {
      error:
        'This book has too little readable text — it is likely an image-only picture book.',
    };
  }

  return { title: title || 'Imported book', text: fullText, chapters };
}

/**
 * Download an EPUB CORS-free and return its bytes, or null when unreachable /
 * empty. The 20 MB cap mirrors the server's `safe_get` EPUB limit.
 */
export async function fetchEpub(url: string): Promise<Uint8Array | null> {
  const res = await corsFreeGetBytes(url, {
    accept: 'application/epub+zip',
    timeoutMs: 30000,
  });
  if (!res.ok || res.bytes.length === 0 || res.bytes.length > 20_000_000) {
    return null;
  }
  return res.bytes;
}
